resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryContext.java

```diff
@@ -342,9 +358,229 @@ final boolean finishFailure(String message, Throwable cause, boolean timeout) {
             // This was caused by a timeout so use DnsNameResolverTimeoutException to allow the user to
             // handle it special (like retry the query).
             e = new DnsNameResolverTimeoutException(nameServerAddr, question, buf.toString());
+            if (retryWithTCP(e)) {
+                // We did successfully retry with TCP.
+                return false;
+            }
         } else {
             e = new DnsNameResolverException(nameServerAddr, question, buf.toString(), cause);
         }
         return promise.tryFailure(e);
     }
+
+    /**
+     * Retry the original query with TCP if possible.
+     *
+     * @param originalResult    the result of the original {@link DnsQueryContext}.
+     * @return                  {@code true} if retry via TCP is supported and so the ownership of
+     *                          {@code originalResult} was transferred, {@code false} otherwise.
+     */
+    private boolean retryWithTCP(final Object originalResult) {
```

Comment:
Can you please elaborate on the motivation in a bit more details? Where this is coming from?
In my experience, DNS timeouts are quite frequent and automatically retrying them via TCP may cause unexpected side-effects, like:
1. People expecting the query to fail within a pre-defined timeout. Retrying it will double the total timeout.
2. Spike of TCP traffic for DNS when timeouts are frequent enough.

I will propose to make this an opt-in feature, but mostly looking for a motivational story behind it.

Reply:
@idelpivnitskiy you may read this https://www.weave.works/blog/racy-conntrack-and-dns-lookup-timeouts 

Dns through tcp is real in needed, maybe is not default behavior, we can let people enable  tcp fallback option and set tcp timeout. 

We use redisson watch dns change event, redisson dns watching implement by netty dns resolver, redisson dns udp timeout 5000ms frequently becase of udp dns race. https://github.com/redisson/redisson/issues/5137

Reply:
Thanks for the link and additional context.

My arguments (1) and (2) remain valid. If we merge this PR, it changes the current behavior and may cause unexpected outages under heavy load and scale. So, if we proceed with this change, it should be configurable (preferable to be opt-in by default) to let impacted users experiment in a safe fashion without affecting everyone.

After reading the article, I'm not sure this specific change is a good mitigation for the described problem. DNS timeouts are quite often, happen not only in k8s env, and may be caused by various different reasons. There are cases, when automatic retry over TCP can make everything even worse. 
Good mitigation points from the article are:
a. disable parallel lookups
b. use TCP for lookups

With (b) and [RFC7766](https://www.rfc-editor.org/rfc/rfc7766) together, it makes more sense to let users chose TCP as the default protocol for their DNS client and skip UDP completely or let it be a fallback. However, it opens other questions around connection management and DoS protection described in RFC7766.

Reply:
We cannot assume what kind of network our program will run on. In any case, a timeout will occur. We should allow users to configure: when a failure occurs, switch to another network connection method.

Reply:
@idelpivnitskiy the problem is that there is really no way for people to to do a retry via TCP by themself (we might want to support also create a `DnsNameResolver` which only does TCP in the future). That said let's just make it configurable to also fallback to TCP on timeout.